python/dgl/sparse.py

```diff
@@ -179,6 +180,86 @@ def _gspmm(gidx, op, reduce_op, u, e):
     return v, (arg_u, arg_e)
 
 
+def _gspmm_hetero(g, op, reduce_op, u_and_e_tuple):
+    r""" Generalized Sparse Matrix Multiplication interface.
+    """
+    num_ntypes = g._graph.number_of_ntypes()
+    u_tuple, e_tuple = u_and_e_tuple[:num_ntypes], u_and_e_tuple[num_ntypes:]
+    gidx = g._graph
+    use_u = op != 'copy_rhs'
+    use_e = op != 'copy_lhs'
+
+    # TODO (Israt): Add check - F.dtype(u) != F.dtype(e):
+
+    # deal with scalar features.
+    expand_u, expand_e = False, False
+    list_u = [None] * gidx.number_of_ntypes()
+    list_v = [None] * gidx.number_of_ntypes()
+    list_e = [None] * gidx.number_of_etypes()
+
+    for srctype, etype, dsttype in g.canonical_etypes:
+        etid = g.get_etype_id(etype)
+        src_id = g.get_ntype_id(srctype)
+        dst_id = g.get_ntype_id(dsttype)
+        u = u_tuple[src_id] if use_u else None
+        e = e_tuple[etid] if use_e else None
+        if use_u:
+            if u is not None and F.ndim(u) == 1:
+                u = F.unsqueeze(u, -1)
+                expand_u = True
+        if use_e:
+            if e is not None and F.ndim(e) == 1:
+                e = F.unsqueeze(e, -1)
+                expand_e = True
+        ctx = F.context(u) if use_u else F.context(e) # TODO(Israt): Put outside of loop
+        dtype = F.dtype(u) if use_u else F.dtype(e) # TODO(Israt): Put outside of loop
+        u_shp = F.shape(u) if use_u else (0,)
+        e_shp = F.shape(e) if use_e else (0,)
+        if use_u:
+            list_u[src_id] = u if use_u else None
+        if use_e:
+            list_e[etid] = e if use_e else None
```

Comment:
why don't you put everything in one `if use_u` condition and `if use_e` condition?